### ISSUE TITLE
[no-unnecessary-condition] false positive on assignment in callbacks such as of Array.map

### ISSUE THREAD

**Issue by FritsvanCampen** [typescript-eslint/typescript-eslint#4315]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/main/docs/linting/TROUBLESHOOTING.md) and my problem is not listed.

**Repro**

```TS
let bool = true;
[ "abc" ].map(() => {
	bool = false;
});
if (bool) { // <-- "ESLint: value is always truthy" - which is obviously not correct
	alert("Can't get me");
}
```

I'm not sure why the rule thinks `bool` is always `true`. Does it not look inside functions? Does it assume the function is not immediately called? Is this rule made so that it skews in favor of false-positives?

**Expected Result**

No warning.

**Actual Result**

A warning.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.7.0` |
| `@typescript-eslint/parser`        | `5.7.0` |
| `TypeScript`                       | `4.5.4` |
| `ESLint`                           | `8.4.1` |
| `node`                             | `v16.4.0` |

**Comment by bradzacher** [comment]
This is not a false positive.
TS cannot understand the control-flow here, so it reports `bool` as always `true`.
https://www.typescriptlang.org/play?#code/DYUwLgBARg9jwQLwTAJwK4gNwCgDaEARAIZQDGhEAugHQC2xADgBTMCUSAfBAN44QDocBMgBmxYAGdsOAL5tcAS1ERmseBx4QA9NogAeALSGiAUQDKAGUUA7MAC4IANwmYIiyRAkB3YgE9PNHQwAAs-ShNvEMUyEPdPGCgnRRh0SWA-CBsYSDIYVFQQMjAcXQEAPQB+fkEJEFQwZkIAYWIbAHJIAHNwCDoQQgU5IA

The rule is driven by TS types - so the rule is only as good as the types you provide.
